clang-tools-extra/clang-tidy/bugprone/TaggedUnionMemberCountCheck.cpp

```diff
@@ -0,0 +1,214 @@
+          has(fieldDecl(hasType(qualType(hasCanonicalType(enumType()))))
+                  .bind("tags")))
+          .bind("root"),
+      this);
+}
+
+static bool isUnion(const FieldDecl *R) {
+  return R->getType().getCanonicalType().getTypePtr()->isUnionType();
+}
+
+static bool isEnum(const FieldDecl *R) {
+  return R->getType().getCanonicalType().getTypePtr()->isEnumeralType();
+}
+
+static bool hasMultipleUnionsOrEnums(const RecordDecl *Rec) {
+  return llvm::count_if(Rec->fields(), isUnion) > 1 ||
+         llvm::count_if(Rec->fields(), isEnum) > 1;
+}
+
+static bool signEquals(const llvm::APSInt &A, const llvm::APSInt &B) {
+  return (A.isNegative() && B.isNegative()) ||
+         (A.isStrictlyPositive() && B.isStrictlyPositive()) ||
+         (A.isZero() && B.isZero());
+}
+
+static bool greaterBySign(const llvm::APSInt &A, const llvm::APSInt &B) {
+  return (A.isNonNegative() && B.isNegative()) ||
+         (A.isStrictlyPositive() && B.isNonPositive());
+}
+
+bool TaggedUnionMemberCountCheck::isCountingEnumLikeName(
+    StringRef Name) const noexcept {
+  if (llvm::any_of(ParsedCountingEnumPrefixes,
+                   [&Name](const StringRef &Prefix) -> bool {
+                     return Name.starts_with_insensitive(Prefix);
+                   }))
+    return true;
+  if (llvm::any_of(ParsedCountingEnumSuffixes,
+                   [&Name](const StringRef &Suffix) -> bool {
+                     return Name.ends_with_insensitive(Suffix);
+                   }))
+    return true;
+  return false;
+}
+
+size_t TaggedUnionMemberCountCheck::getNumberOfValidEnumValues(
+    const EnumDecl *Ed) noexcept {
+  bool FoundMax = false;
+  llvm::APSInt MaxTagValue;
+  llvm::SmallSet<llvm::APSInt, 32> EnumValues;
+
+  // Heuristic for counter enum constants.
+  //
+  //   enum tag_with_counter {
+  //     tag1,
+  //     tag2,
+  //     tag_count, <-- Searching for these enum constants
+  //   };
+  //
+  // The final tag count is decreased by 1 if and only if:
+  // 1. There is only one counting enum constant,
+  // 2. The counting enum constant is the last enum constant that is defined,
+  // 3. The value of the counting enum constant is the largest out of every enum
+  //    constant.
+  // The 'ce' prefix is a shorthand for 'counting enum'.
+  size_t CeCount = 0;
+  bool CeIsLast = false;
+  llvm::APSInt CeValue = llvm::APSInt::get(0);
+
+  for (const auto &Enumerator : Ed->enumerators()) {
+    const llvm::APSInt Val = Enumerator->getInitVal();
+    EnumValues.insert(Val);
+    if (FoundMax) {
+      if (greaterBySign(Val, MaxTagValue) ||
+          (signEquals(Val, MaxTagValue) && Val > MaxTagValue))
+        MaxTagValue = Val;
+    } else {
+      MaxTagValue = Val;
+      FoundMax = true;
+    }
+
+    if (EnableCountingEnumHeuristic) {
+      if (isCountingEnumLikeName(Enumerator->getName())) {
+        CeIsLast = true;
+        CeValue = Val;
+        CeCount += 1;
+        CountingEnumConstantDecl = Enumerator;
```

Comment:
This `CountingEnumConstantDecl` is a quasi-global symbol as the member of the check class. Also, it is not cleared after the last diagnostic is printed, and we only rely on the class never getting used again in order to not have something dangling here.

Perhaps it would be better if the function returned `std::pair<std::size_t, const Enum...*>`, or a `const Enum...*&` output parameter was taken as a side effect.